ta/pkcs11/src/pkcs11_token.c

```diff
@@ -1464,3 +1464,85 @@ enum pkcs11_rc entry_ck_logout(struct pkcs11_client *client,
 
 	return PKCS11_CKR_OK;
 }
+
+enum pkcs11_rc entry_ck_seed_random(struct pkcs11_client *client,
+				    uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *in = params + 1;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (in->memref.size && !in->memref.buffer)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	/* noop - just ignore the input seed */
+
+	IMSG("PKCS11 session %"PRIu32": seed random", session->handle);
+
+	return PKCS11_CKR_OK;
+}
+
+enum pkcs11_rc entry_ck_generate_random(struct pkcs11_client *client,
+					uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	void *buffer = NULL;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (out->memref.size && !out->memref.buffer)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (!out->memref.size)
+		return PKCS11_CKR_OK;
+
+	buffer = TEE_Malloc(out->memref.size, TEE_MALLOC_FILL_ZERO);
+	if (!buffer)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	TEE_GenerateRandom(buffer, out->memref.size);
```

Comment:
I did not understand why trusted_keys's TEE_GenerateRandom() doesn't require TEE_Malloc():
https://github.com/OP-TEE/optee_os/blob/master/ta/trusted_keys/entry.c#L45

Tried to observe from GlobalPlatform specification and did not notice why there is required to have ACL check in TEE_GenerateRandom(). So should that be relaxed a bit or does trusted_keys also need fixing?

Reply:
`TEE_GenerateRandom()` has a `[out]` annotation on the `randomBuffer` pointer which means (among other things) that it must reside in shared memory, in other words it must be private to the TA.

Reply:
OK found the annotation description.

Reply:
Just a note about trusted_keys -- Linux kernel has optee-rng driver which when configured should provide additional seed to Linux kernel's already good RNG -- so I don't understand why there is separate entry point for this -- but I leave this to others to resolve.

Reply:
@jforissier

> So yes I would think trusted_keys needs fixing. How is it tested? It seems not :-/

Its tested but I don't think its tested after this patch [1] which has switched to stricter memory requirements for `randomBuffer` to reside in TA private memory only. So will fix it using `TEE_Malloc()` instead.

[1] https://github.com/OP-TEE/optee_os/commit/e12c9f67d12c24a39ec0b98049e445bb3ae7a23e

@vesajaaskelainen Trusted Keys in Linux are meant to be generated, sealed and unsealed via trust source only. See this [2] discussion for reference.

[2] https://lore.kernel.org/linux-integrity/CAE=NcrY3BTvD-L2XP6bsO=9oAJLtSD0wYpUymVkAGAnYObsPzQ@mail.gmail.com/T/#t

Reply:
@b49020 thanks for following up on this. Note, when I wrote it seemed not tested I didn't mean to say you tested poorly :wink: but that we don't have a systematic test in `optee_test` for instance. Which would be nice to have one all the dependencies are stable.

Reply:
I agree to have a systematic test in `optee_test`. See related discussion here [1].

[1] https://github.com/OP-TEE/optee_test/issues/488